core/src/Streamly/Internal/Data/Serialize.hs

```diff
@@ -0,0 +1,152 @@
+module Streamly.Internal.Data.Serialize
+    ( Size(..)
+    , Serialize(..)
+    ) where
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Control.Monad (void)
+import Data.List (foldl')
+import Data.Proxy (Proxy (..))
+import Streamly.Internal.Data.Unboxed (MutableByteArray(..))
+
+import GHC.Int (Int16(..), Int32(..), Int64(..), Int8(..))
+import GHC.Word (Word16(..), Word32(..), Word64(..), Word8(..))
+import GHC.Stable (StablePtr(..))
+
+import qualified Streamly.Internal.Data.Unboxed as Unbox
+
+import GHC.Exts
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+-- | Info about the length of a serializable type. Length can depend on the
+-- value or can be independent.
+data Size a
+    = VarSize (a -> Int)
+    | ConstSize !Int
+
+-- | A type implementing the 'Serialize' interface supplies operations for
+-- reading and writing the type from and to a mutable byte array (an unboxed
+-- representation of the type) in memory. The read operation 'deserialize'
+-- deserializes the boxed type from the mutable byte array. The write operation
+-- 'serialize' serializes the boxed type to the mutable byte array.
```

Comment:
We should mention that the byte ordering in the serialized data is the same as the machine byte order. Therefore, it cannot be compatible across machines with different byte ordering.